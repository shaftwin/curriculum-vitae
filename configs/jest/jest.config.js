const path = require('path');

module.exports = {
  rootDir: '../../',
  testEnvironment: 'jsdom',
  testMatch: ['**/*.test.js'],
  setupFilesAfterEnv: ['./node_modules/@testing-library/jest-dom'],
  moduleDirectories: [path.resolve(__dirname, '../../'), 'node_modules'],
  setupFiles: [
    './configs/jest/mock.js',
    './configs/cucumber/jest-cucumber-config.js',
  ],
  snapshotSerializers: ['@emotion/jest/serializer'],
};
