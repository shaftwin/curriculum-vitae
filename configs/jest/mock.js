jest.mock('react-i18next', () => ({
  withTranslation: () => (Component) => {
    // eslint-disable-next-line no-param-reassign
    Component.defaultProps = { ...Component.defaultProps, t: () => '' };
    return Component;
  },
  // this mock makes sure any components using the
  // translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));
