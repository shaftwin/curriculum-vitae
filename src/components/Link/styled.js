import styled from '@emotion/styled';

export const Container = styled.a`
  text-decoration: none;
  color: ${({ theme, templateColor }) =>
    theme.link.textColor[templateColor].default};
  font-size: ${({ theme, size }) => theme.link.fontSize[size]};
  font-weight: ${({ theme, size }) => theme.link.fontWeight[size]};

  &:hover {
    color: ${({ theme, templateColor }) =>
      theme.link.textColor[templateColor].hover};
  }

  &:visited: {
    color: ${({ theme, templateColor }) =>
      theme.link.textColor[templateColor].visited};
  }
`;
