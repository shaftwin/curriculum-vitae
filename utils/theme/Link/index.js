import { global } from '../global';

export const link = {
  textColor: {
    primary: {
      default: global.colors.green,
      hover: global.colors.hover.green,
      visited: global.colors.mediumGrey,
    },
    secondary: {
      default: global.colors.darkBlue,
      hover: global.colors.hover.green,
      visited: global.colors.mediumGrey,
    },
    white: {
      default: global.colors.white,
      hover: global.colors.hover.green,
      visited: global.colors.disabledGrey,
    },
  },
  fontSize: {
    extraSmall: global.fontSize[11],
    small: global.fontSize[12],
    normal: global.fontSize[13],
  },
  fontWeight: {
    extraSmall: global.fontWeight.semiBold,
    small: global.fontWeight.semiBold,
    normal: global.fontWeight.bold,
  },
};
