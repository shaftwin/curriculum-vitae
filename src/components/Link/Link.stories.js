/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import styled from '@emotion/styled';
import Link from './index';

export default {
  title: 'Components/Link',
  component: Link,
};

const GreenContainer = styled.div`
  background-color: ${({ theme }) => theme.global.colors.green};
  width: 12.5rem;
`;

const Template = (args) => <Link {...args} />;

const WhiteTemplate = (args) => (
  <GreenContainer>
    <Link {...args} />
  </GreenContainer>
);

export const Default = Template.bind({});
Default.args = {
  text: 'Link',
  link: 'https://www.google.com/',
};

export const WhiteWithGreenBackground = WhiteTemplate.bind({});
WhiteWithGreenBackground.args = {
  text: 'White Link',
  templateColor: 'white',
  link: 'https://www.google.com/',
};
