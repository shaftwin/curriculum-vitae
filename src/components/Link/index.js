import React from 'react';
import PropTypes from 'prop-types';
import { Container } from './styled';

const Link = ({ className, text, link, templateColor, size, ...props }) => {
  return (
    <Container
      className={className}
      href={link}
      templateColor={templateColor}
      size={size}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
    >
      {text}
    </Container>
  );
};

Link.propTypes = {
  /**
   * Required text
   */
  text: PropTypes.string.isRequired,
  /**
   * Required link
   */
  link: PropTypes.string.isRequired,
  /**
   * Optional template color
   */
  templateColor: PropTypes.oneOf(['primary', 'secondary', 'white']),
  /**
   * Optional size
   */
  size: PropTypes.oneOf(['extraSmall', 'small', 'normal']),
  /**
   * Optional custom style
   */
  className: PropTypes.string,
};

Link.defaultProps = {
  templateColor: 'primary',
  size: 'normal',
};

export default Link;
