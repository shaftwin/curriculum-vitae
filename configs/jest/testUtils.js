// import { MockedProvider } from '@apollo/client/testing';
import { ThemeProvider } from '@emotion/react';
import { render } from '@testing-library/react';
import React from 'react';
import theme from '../../utils/theme';

const AllTheProviders = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
