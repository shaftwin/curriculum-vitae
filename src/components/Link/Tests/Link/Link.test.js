import * as React from 'react';
import { screen, waitForElementToBeRemoved } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { defineFeature, loadFeature } from 'jest-cucumber';
import { render } from '../../../../../configs/jest/testUtils';
import Link from '../..';

const feature = loadFeature('src/components/Link/Tests/Link/Link.feature');

defineFeature(feature, (test) => {
  let container;

  // TODO: Externalize this method
  const givenLinkIsRendered = (given, comp) => {
    given('Link is mounted', () => {
      container = render(comp).container;
    });
  };

  test('Default link', ({ given, then, and }) => {
    givenLinkIsRendered(
      given,
      <Link text="testLink" link="https://www.google.com/" />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByText(/testLink/i)).toBeInTheDocument();
    });

    then('Link should have the default style', () => {
      expect(container).toMatchSnapshot();
    });
  });

  test('Link hovered', ({ given, when, then, and }) => {
    givenLinkIsRendered(
      given,
      <Link text="testLink" link="https://www.google.com/" />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByText(/testLink/i)).toBeInTheDocument();
    });

    when('Link is hovered', () => {
      userEvent.hover(screen.queryByText(/testLink/i));
    });

    then('Link should have the hover style', () => {
      expect(container).toMatchSnapshot();
    });
  });

  test('Link clicked', ({ given, when, then, and }) => {
    givenLinkIsRendered(
      given,
      <Link text="testLink" link="https://www.google.com/" />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByText(/testLink/i)).toBeInTheDocument();
    });

    when('Click event is fired', () => {
      userEvent.click(screen.getByText(/testLink/i));
    });

    then('Link should have the visited style', () => {
      expect(container).toMatchSnapshot();
    });

    and('The user should be redirected to the right page', () => {
      waitForElementToBeRemoved(screen.queryByText(/testLink/i))
        .then(() => {
          expect(screen.queryByText(/testLink/i)).not.toBeInTheDocument();
        })
        .catch(() => {
          // eslint-disable-next-line no-console
          console.log('User not redirected');
        });
    });
  });
});
