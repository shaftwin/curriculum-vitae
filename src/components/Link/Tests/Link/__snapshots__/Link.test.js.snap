// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Default link 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #00DBC7;
  font-size: 0.813rem;
  font-weight: 700;
}

.emotion-0:hover {
  color: #00C6B3;
}

.emotion-0:visited: {
  color: #A4AAC0;
}

<div>
  <a
    class="emotion-0"
    href="https://www.google.com/"
  >
    testLink
  </a>
</div>
`;

exports[`Link Link clicked 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #00DBC7;
  font-size: 0.813rem;
  font-weight: 700;
}

.emotion-0:hover {
  color: #00C6B3;
}

.emotion-0:visited: {
  color: #A4AAC0;
}

<div>
  <a
    class="emotion-0"
    href="https://www.google.com/"
  >
    testLink
  </a>
</div>
`;

exports[`Link Link hovered 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #00DBC7;
  font-size: 0.813rem;
  font-weight: 700;
}

.emotion-0:hover {
  color: #00C6B3;
}

.emotion-0:visited: {
  color: #A4AAC0;
}

<div>
  <a
    class="emotion-0"
    href="https://www.google.com/"
  >
    testLink
  </a>
</div>
`;
